core/client/fs/src/main/java/alluxio/client/block/stream/NaiveCachedGrpcDataReader.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.DataMessage;
+import alluxio.grpc.ReadRequest;
+import alluxio.grpc.ReadResponse;
+import alluxio.grpc.ReadResponseMarshaller;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.resource.CloseableResource;
+import alluxio.wire.WorkerNetAddress;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import alluxio.resource.LockResource;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import com.google.protobuf.ByteString;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A gRPC data reader that streams a region from gRPC data server.
+ *
+ * Protocol:
+ * 1. The client sends a read request (id, offset, length).
+ * 2. Once the server receives the request, it streams chunks to the client. The streaming pauses
+ *    if the server's buffer is full and resumes if the buffer is not full.
```

Comment:
if the client's buffer is full?

Reply:
Will we add a mechanism to limit the available buffer used for the NaiveReader and make the available buffer configurable? Can be done in future PRs.
Kind of worry about the data reader introduces the client-side OOM issue.